ngtpn - Technical Training

nguyenntph opened the pull request.

ngtpn committed:
[ADD] estate: Add module

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1024.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1024)

ngtpn committed:
[IMP] estate: Add name / expected price to property model

ngtpn committed:
[IMP] estate: Add access right

ngtpn committed:
[IMP] estate: Add UIs / Update fields

clbr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,45 @@
+from odoo import fields, models
+from dateutil.relativedelta import relativedelta
```
Please follow our [coding guidelines](https://www.odoo.com/documentation/19.0/contributing/development/coding_guidelines.html#imports) regarding order of imports.

clbr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,45 @@
+from dateutil.relativedelta import relativedelta
+
+
+class Property(models.Model):
```
```suggestion
class EstateProperty(models.Model):
```
It won't have big incidence here but we generally put the name of the file here, in CamelCase.

nguyenntph replied:
This is now updated. I just want to check if we have a convention for this. As in our [Chapter 3](https://www.odoo.com/documentation/19.0/developer/tutorials/server_framework_101/03_basicmodel.html) example [CRM Recurring Plan](https://github.com/odoo/odoo/blob/e80911aaead031e7523173789e946ac1fd27c7dc/addons/crm/models/crm_recurring_plan.py#L1-L9), we omit the add on name and just use the model for class name, so I was inferring that we have a specific naming convention, like add-on-name_model-name for file, and model-name only for class name.

clbr-odoo replied:
No it's not a strong official convention, more an employee tacit thing, and it's indeed not always the case.
Always remember that some part of the Odoo code are 10 years old, when there was 20 devs :wink:

clbr-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,45 @@
+    postcode = fields.Char()
+    date_availability = fields.Date(
+        copy=False,
+        default=fields.Datetime.today() + relativedelta(months=+3),
```
```suggestion
        default=fields.Date.add(fields.Date.today(), months=3),
```
1. Your code is (almost, see 2) correct, just know that we have some built-in helper for that.
It allows to avoid the import of relativedelta :)
2. You should use `fields.Date.today()` instead of Datetime since you have a `Date' type field. Otherwise you trigger an extra cast needlessly.

nguyenntph replied:
Thanks for that. I was looking into the helper but missed this `fields.Date.add()`. This is now updated.

clbr-odoo commented on estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,7 @@
+<odoo>
+  <menuitem id="test_menu_root" name="Real Estate">
```
I don't think there is a generic "Odoo rule" for that, but in my team we always indent with 4 spaces, even in XMLs.
And I guess the `test_` is not really necessary in those `xmlids`, otherwise you'll end up prepending `test` everywhere in this tutorial 😆

nguyenntph replied:
Good point. I notice I used mixed 2 and 4 spaces. This is now updated.

clbr-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,16 @@
+    'license': 'LGPL-3',
+    'version': '1.0',
+    'depends': ['base'],
+    'author': "ngtpn",
```
No problem in this tutorial, just know that when developping an Odoo official module, we must use `'author': 'Odoo S.A.',`.

nguyenntph replied:
Good point. Thanks for sharing that.

clbr-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,16 @@
+    'description': """
+    Real Estate Advertisement module
+    """,
+    'data': [
```
```suggestion
    'application': True,
    'data': [
```
You probably want to have application=True for this module. See the [manifest reference](https://www.odoo.com/documentation/19.0/developer/reference/backend/module.html)

clbr-odoo reviewed (changes requested):
Nice to see that you know of to rewrite your git history & force-push.
I advice to use the `--force-with-lease` option if you haven't, as a general security.

Very good job so far 👍 
Props for the clean indentation of the Python file 🥳

ngtpn committed:
[IMP] estate: Fix xml indentation

ngtpn committed:
[IMP] estate: Use date helper / Rename model

ngtpn committed:
[IMP] estate: Add list / form / search UI

ngtpn committed:
[IMP] estate: Add perperty type / tag / offer

ngtpn committed:
[IMP] estate: Compute / Inverse field value

ngtpn committed:
[IMP] estate: Add actions

ngtpn committed:
[IMP] estate: Add validation conditions

ngtpn committed:
[IMP] estate: Sprinkes

- Inline views
- Statusbar
- Sort order
- Field options
- Readonly

ngtpn committed:
[REF] estate: Update view files

ngtpn committed:
[IMP] estate: Sprinkles (cont)

- Color picker
- Editable list view
- Decorators
- Fix no create/edit
- Add offer link
- Add domain

ngtpn committed:
[IMP] estate: Inheritance

ngtpn committed:
[ADD] estate_account: Init

ngtpn committed:
[IMP] estate_account: Create account move on sold properties

ngtpn committed:
[IMP] estate: Kanban view

yoba-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,120 @@
+    @api.depends('offer_ids')
+    def _compute_best_price(self):
+        for record in self:
+            record.best_price = max(record.offer_ids.mapped('price')) if self.offer_ids else 0
```
```suggestion
            record.best_price = max(record.offer_ids.mapped('price'), default=0.0)
```

yoba-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,120 @@
+
+class EstateProperty(models.Model):
+    _name = 'estate.property'
+    _description = 'All properties'
```
```suggestion
    _description = 'Estate Properties'
```
Nitpick :innocent: 
Also description should always be in title case

nguyenntph replied:
Good point. All updated for all models :smiling_face_with_tear:

yoba-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,120 @@
+            self.garden_area = 10
+            self.garden_orientation = 'north'
+        else:
+            self.garden_area = None
```
```suggestion
            self.garden_area = 0
```

yoba-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,71 @@
+
+class EstatePropertyOffer(models.Model):
+    _name = 'estate.property.offer'
+    _description = 'All offers'
```
Same as estate property :slightly_smiling_face:

yoba-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,71 @@
+    partner_id = fields.Many2one('res.partner', string='Partner')
+    property_id = fields.Many2one('estate.property', string='Property')
+    validity = fields.Integer(default=7)
+    date_deadline = fields.Date(compute='_compute_deadline', inverse="_inverse_deadline")
```
```suggestion
    date_deadline = fields.Date(compute='_compute_date_deadline', inverse="_inverse_date_deadline")
```
When comes to naming the compute and inverse method it is always better to use the whole field name

yoba-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,121 @@
+    def action_mark_as_sold(self):
+        for record in self:
+            if record.state == 'cancelled':
+                raise UserError('Property is already cancelled')
```
```suggestion
                raise UserError('Cannot sell cancelled properties')
```
For errors it is better to be precise about why can't we do the action not just that we can't do it

nguyenntph replied:
Thank you. That's a good point. Error messages updated.

yoba-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,121 @@
+    def action_mark_as_cancelled(self):
+        for record in self:
+            if record.state == 'sold':
+                raise UserError('Property is already sold')
```
Same :point_up: :smile:

yoba-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,71 @@
+                created_date, days=record.validity,
+            )
+
+    @api.model
+    def create(self, val_lists):
+        for vals in val_lists:
+            linked_property = self.env['estate.property'].browse(vals['property_id'])
+            if vals['price'] < linked_property.best_price:
+                raise UserError('Offer price cannot be lower than existing offers')
+
+            linked_property.state = 'offer_received'
+        return super().create(val_lists)
```
We have guidelines also about the order of the methods in a file you can check it here https://www.odoo.com/documentation/19.0/contributing/development/coding_guidelines.html#symbols-and-conventions

Also when we have a compute and inverse method for a field we tend to put the compute then right after it the inverse

yoba-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,71 @@
+    def action_mark_as_accepted(self):
+        for record in self:
+            if (
+                record.property_id.state != 'new'
+                and record.property_id.state != 'offer_received'
+            ):
```
```suggestion
            if (
                record.property_id.state not in ('new', 'offer_received')
            ):
```

yoba-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,71 @@
+        for record in self:
+            if record.status == 'accepted':
+                record.property_id.state = 'offer_received'
+                record.property_id.selling_price = None
```
```suggestion
                record.property_id.selling_price = 0.0
```

yoba-odoo commented on estate/models/estate_property_tag.py:
```diff
@@ -0,0 +1,15 @@
+
+class EstatePropertyTag(models.Model):
+    _name = 'estate.property.tag'
+    _description = 'All property tags'
```
Surprise!! Same as above :smile:

yoba-odoo commented on estate/models/estate_property_type.py:
```diff
@@ -0,0 +1,22 @@
+
+class EstatePropertyType(models.Model):
+    _name = 'estate.property.type'
+    _description = 'All property types'
```
You know already :sweat_smile:

yoba-odoo commented on estate/views/estate_property_offer_views.xml:
```diff
@@ -0,0 +1,71 @@
+                        </group>
+                        <group>
+                            <field name="date_deadline" string="Deadline"/>
+                        </group>
```
You don't have to put each field alone in a group. Am I missing something?

nguyenntph replied:
Seems like we don't. This is the code block we used as example in [Chapter 6](https://www.odoo.com/documentation/19.0/developer/tutorials/server_framework_101/06_basicviews.html#form), using group for each fields. 

<img width="1039" height="566" alt="Codeblock" src="https://github.com/user-attachments/assets/7f2d5c5f-5604-4039-b133-0552f49f2a78" />

I tried having the fields in 1 group and everything works fine.

yoba-odoo commented on estate/views/estate_property_tag_views.xml:
```diff
@@ -0,0 +1,35 @@
+                <sheet>
+                    <group>
+                        <field name="name"/>
+                    </group>
+                    <group>
+                        <field name="color" widget="color_picker"/>
+                    </group>
```
same

nguyenntph replied:
Same as above as well.

yoba-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,191 @@
+
+                <separator/>
+
+                <filter string="Available" name="available" domain="['|', ('state', '=', 'new'), ('state', '=', 'offer_received')]"/>
```
```suggestion
                <filter string="Available" name="available" domain="[('state', 'in', ('new', 'offer_received'))]"/>
```

yoba-odoo reviewed (changes requested):
Nice work :clap: 
Left some non-blocking comments

ngtpn committed:
[REF] estate: Refactor based on guidelines and feedbacks

- Model description
- Error messages
- Views group
- Default data type
- Simplify conditional check

clbr-odoo committed:
[FIX] awesome_owl: Fix missing css variables

Some imports are missing in the manifest causing some warnings and
the css doesn't load properly.
Fix was fixed in the master branch, backporting it in 19.0
for the onboarding classes that always happen in the lastest stable.

task-none

Part-of: odoo/tutorials#1037
Signed-off-by: Antoine Vandevenne (anv) <anv@odoo.com>

clbr-odoo committed:
[FIX] awesome_dashboard: Get rid of the deprecated warning

`json` routes were deprecated to `jsonrpc` in 19.0, let's
get rid of the warning to avoid confusion for the newdoos.

task-none

closes odoo/tutorials#1037

Signed-off-by: Antoine Vandevenne (anv) <anv@odoo.com>

ngtpn committed:
[IMP] awesome_owl: Add counter / card / sum

ngtpn committed:
[IMP] awesome_owl: Add todo list

ngtpn committed:
[IMP] awesome_owl: Add todo item

ngtpn committed:
[IMP] awesome_owl: Auto focus

ngtpn committed:
[IMP] awesome_owl: Toggle / Delete items

ngtpn committed:
[IMP] awesome_owl: Update styles

ngtpn committed:
[IMP] awesome_owl: Card slot / toggle

ngtpn committed:
[IMP] awesome_dashboard: Add Layout / navigation buttons

ngtpn committed:
[IMP] awesome_dashboard: Dashboard item

ngtpn committed:
[IMP] awesome_dashboard: Show statistics

ngtpn committed:
[IMP] awesome_dashboard: statistics service

ngtpn committed:
[IMP] awesome_dashboard: Add pie chart

ngtpn committed:
[IMP] awesome_dashboard: real life update

ngtpn committed:
[IMP] awesome_dashboard: Lazy load dashboard

ngtpn committed:
[IMP] awesome_dashboard: Generic dashboard

ngtpn committed:
[IMP] awesome_dashboard: Fix reload

ngtpn committed:
[IMP] awesome_dashboard: Extensible dashboard

ngtpn committed:
[IMP] awesome_dashboard: Allow selecting items in dashboard
